api/v2/performanceprofile_validation.go

```diff
@@ -0,0 +1,277 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+const (
+	hugepagesSize2M = "2M"
+	hugepagesSize1G = "1G"
+)
+
+var _ webhook.Validator = &PerformanceProfile{}
+
+// we need this variable only because our validate methods should have access to the client
+var validatorClient client.Client
+
+// SetupWebhookWithManager enables Webhooks - needed for version conversion
+func (r *PerformanceProfile) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	if validatorClient == nil {
+		validatorClient = mgr.GetClient()
+	}
+
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *PerformanceProfile) ValidateCreate() error {
+	klog.Infof("Create validation for the performance profile %q", r.Name)
+
+	return r.validateCreateOrUpdate()
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *PerformanceProfile) ValidateUpdate(old runtime.Object) error {
+	klog.Infof("Update validation for the performance profile %q", r.Name)
+
+	return r.validateCreateOrUpdate()
+}
+
+func (r *PerformanceProfile) validateCreateOrUpdate() error {
+	var allErrs field.ErrorList
+
+	// validate node selector duplication
+	ppList := &PerformanceProfileList{}
+	if err := validatorClient.List(context.TODO(), ppList); err != nil {
+		allErrs = append(allErrs, field.InternalError(field.NewPath("spec.nodeSelector"), err))
+		return apierrors.NewInternalError(err)
+	}
+
+	allErrs = append(allErrs, r.validateNodeSelectorDuplication(ppList)...)
+
+	// validate basic fields
+	allErrs = append(allErrs, r.validateFields()...)
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "performance.openshift.io", Kind: "PerformanceProfile"},
+		r.Name, allErrs)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *PerformanceProfile) ValidateDelete() error {
+	klog.Infof("Delete validation for the performance profile %q", r.Name)
+
+	// TODO(user): fill in your validation logic upon object deletion.
+	return nil
+}
+
+func (r *PerformanceProfile) validateNodeSelectorDuplication(ppList *PerformanceProfileList) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// validate node selector duplication
+	for _, pp := range ppList.Items {
+		// exclude the current profile from the check
+		if pp.Name == r.Name {
+			continue
+		}
+
+		// node selectors have different length and it can work on the same set of nodes
+		if reflect.DeepEqual(pp.Spec.NodeSelector, r.Spec.NodeSelector) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec.nodeSelector"), r.Spec.NodeSelector, fmt.Sprintf("the profile has the same node selector as the performance profile %q", pp.Name)))
+		}
+	}
+
+	return allErrs
+}
+
+func (r *PerformanceProfile) validateFields() field.ErrorList {
```

Comment:
are we turning what formerly where free functions into methods, right? What's the benefit of this specific move?
Moving the validation into the webhook is fine and welcome, I'm interested about functions vs methods

Reply:
The controller-runtime requires that the object should implement a validation interface to activate the validation webhook for it, so upper methods should be methods. And instead of passing the object, I changed all other functions to methods as well, IMHO it placed under the same file and used only for validation so it no reason to make them functions. If you have some strong objectives I can roll back this change.

Reply:
I don't, both approaches have pros and cons. I don't have strong opinions in either direction so let's go like you did, no point in wasting this work.